File: leetcode/0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
import java.util.*;

class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();

        if (digits == null || digits.length() == 0) return result;

        String[] map = {
            "", "", "abc", "def", "ghi",
            "jkl", "mno", "pqrs", "tuv", "wxyz"
        };

        backtrack(result, digits, map, 0, new StringBuilder());
        return result;
    }

    private void backtrack(List<String> result, String digits, String[] map,
                           int index, StringBuilder path) {

        if (index == digits.length()) {
            result.add(path.toString());
            return;
        }

        String letters = map[digits.charAt(index) - '0'];

        for (char c : letters.toCharArray()) {
            path.append(c);
            backtrack(result, digits, map, index + 1, path);
            path.deleteCharAt(path.length() - 1); // 되돌리기
        }
    }
}

File: leetcode/0017-letter-combinations-of-a-phone-number.ts
function letterCombinations(digits: string): string[] {
    if (!digits || digits.length === 0) return [];

    const map: Record<string, string> = {
        '2': 'abc', '3': 'def', '4': 'ghi',
        '5': 'jkl', '6': 'mno', '7': 'pqrs',
        '8': 'tuv', '9': 'wxyz'
    };

    const result: string[] = [];

    const backtrack = (index: number, path: string): void => {
        if (index === digits.length) {
            result.push(path);
            return;
        }

        const letters = map[digits[index]];

        for (const char of letters) {
            backtrack(index + 1, path + char);
        }
    };

    backtrack(0, "");
    return result;
}